sdk/core/Azure.Core/src/DynamicData/DynamicData.AllowList.cs

```diff
@@ -0,0 +1,217 @@
+            {
+                if (value == null)
+                {
+                    return;
+                }
+
+                Type type = value.GetType();
+
+                if (!IsAllowedType(type, value))
+                {
+                    throw new NotSupportedException($"Type is not currently supported: '{type}'.");
+                }
+            }
+
+            public static bool IsAllowedType<T>(Type type, T value)
+            {
+                if (IsAllowedLeafType(type))
+                {
+                    return true;
+                }
+
+                if (IsAllowedCollectionType(type, value))
+                {
+                    return true;
+                }
+
+                return IsAllowedAnonymousType(type, value, new HashSet<Type>());
+            }
+
+            private static bool IsAllowedLeafType(Type type)
+            {
+                return IsAllowedPrimitive(type) ||
+                    type == typeof(string) ||
+                    type == typeof(DateTime) ||
+                    type == typeof(DateTimeOffset) ||
+                    type == typeof(TimeSpan) ||
+                    type == typeof(Uri) ||
+                    type == typeof(Guid) ||
+                    type == typeof(ETag) ||
+
+                    // Allowable for DynamicData assignments.
+                    type == typeof(JsonElement) ||
+                    type == typeof(JsonDocument) ||
+                    type == typeof(MutableJsonDocument) ||
+                    type == typeof(MutableJsonElement) ||
+                    type == typeof(DynamicData);
+            }
+
+            private static bool IsAllowedPrimitive(Type type)
+            {
+                return
+                    type.IsPrimitive ||
+                    type == typeof(decimal);
+            }
+
+            private static bool IsAllowedCollectionType<T>(Type type, T value)
+            {
+                return
+                    IsAllowedArrayType(type, value) ||
+                    IsAllowedListType(type, value) ||
+                    IsAllowedDictionaryType(type, value);
+            }
+
+            private static bool IsAllowedArrayType<T>(Type type, T value)
+            {
+                if (value is not Array array)
+                {
+                    return false;
+                }
+
+                Type? elementType = type.GetElementType();
+                if (elementType == null)
+                {
+                    return false;
+                }
+
+                if (IsAllowedPrimitive(elementType))
+                {
+                    return true;
+                }
+
+                return IsAllowedEnumerableValue(elementType, array);
+            }
+
+            private static bool IsAllowedListType<T>(Type type, T value)
+            {
+                if (!type.IsGenericType)
+                {
+                    return false;
+                }
+
+                if (value is not IList list)
```

Comment:
Why do we have the check? I liked the code before where we just let List<T>

Reply:
Curious: why did you prefer that?

Without converting it to IList we have to double-cast to IEnumerable in line 124 below.  If there's another way, or you prefer that approach, we can go with that, let me know.

Reply:
Oh, I see, it's more restrictive.  I'll do the double cast.

Reply:
Went back to the `List<T>` approach.